proposals/1125_dynamic-supervisor-oidc-clients/README.md

```diff
@@ -0,0 +1,507 @@
+example, the existing token credential request API could not be implemented using a CRD - it processes an incoming token
+and returns a client certificate without writing any data to etcd.  Aggregated APIs have no such limitation.  We will
+use an aggregated API to handle generation of client secrets.
+
+We will use the existing token credential request API as a model for the new OIDC client secret request API.  Only the
+`create` verb will be supported (but this resource will be part of the `pinniped` category and will have no-op `list`
+implementation just like token credential request to prevent `kubectl get pinniped -A` from returning an error).
+
+```go
+type OIDCClientSecretRequest struct {
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty"`  // metadata.name must be set to the client ID
+
+
+  Spec   OIDCClientSecretRequestSpec   `json:"spec"`
+  Status OIDCClientSecretRequestStatus `json:"status"`
+}
+
+type OIDCClientSecretRequestSpec struct {
+  GenerateNewSecret bool `json:"generateNewSecret"`
+  RevokeOldSecrets  bool `json:"revokeOldSecrets"`
+}
+
+type OIDCClientSecretRequestStatus struct {
+  GeneratedSecret    string `json:"generatedSecret,omitempty"`
+  TotalClientSecrets int    `json:"totalClientSecrets"`
+}
+```
+
+Unlike token credential request, OIDC client secret request will require that `metadata.name` be set (so that it can
+determine what OAuth client is being referred to).  When `.spec.generateNewSecret` is set to `true`, the response will
+provide the plaintext client secret via the `.status.generatedSecret` field.  This is the only time that the plaintext
+client secret is made available.  To aid in rotation, this API may be called multiple times with
+`.spec.generateNewSecret` set to `true` to cause the creation of a new client secret.  The response will include the
+total number of client secrets (including any newly generated ones) that exist for the OAuth client in the
+`.status.totalClientSecrets` field.  When the admin is ready, they may call the API with `.spec.revokeOldSecrets` set to
+`true` to cause all but the latest secret to be revoked.  In the event of a client secret disclosure, a "hard" rotation
+may be performed by setting both `.spec.generateNewSecret` and `.spec.revokeOldSecrets` to `true` (this will revoke all
+pre-existing client secrets and return a newly generated secret).  Leaving both of these fields set to `false` will
+simply return the number of existing client secrets via the `.status.totalClientSecrets` field (this same information is
+available via the `.status.totalClientSecrets` field of the `OIDCClient` resource).
+
+An admin would interact with this API by using standard `kubectl` commands:
+
+```yaml
+apiVersion: oauth.virtual.supervisor.pinniped.dev/v1alpha1  # different group to avoid collision with the CRD
+kind: OIDCClientSecretRequest
+metadata:
+  name: client.oauth.pinniped.dev-j77kz
+  namespace: pinniped-supervisor
```

Comment:
So if I, as a user, need RBAC to `create` an `OIDCClintSecretRequest` in the `pinniped-supervisor` namespace to be able to create new secrets or revoke old ones for my client, is there some finer-grained permissions required to stop me from doing so with other clients? If it's that only an admin with access to all clients can do this (ie. effectively a cluster-admin), then how am I safely passing the resulting client secret to the team that will use it with their web app or similar? I'm wondering if these requests should instead be in my own namespace so that as an admin of my namespace I can get the client secret etc.?

Reply:
That's a good question. These clients are not meant to be self-serve created by the web app developer (yet). They are meant to be managed by the Supervisor admin, so the Supervisor admin would register a client for your app and then send you the credential. Supporting the self-service model is left for future possible consideration.

Reply:
Yeah ok. Still seems to me like the request belongs in the users namespace given that's where the secret would usually end up anyway (I think?), even if it requires an admin to approve by creating some related resource (for now) in the supervisor namespace. The controller could even, once the admin has manually approved the request somehow, create the *secret* resource in the same (user) namespace so that the disappearing status secret below isn't an issue (assuming the secret is going to end up being a secret resource in the user namespace that gets mounted somewhere anyway?). But I'm probably lacking lots of context here, could be many other reasons for the current approach. And perhaps the client secret won't need to be a secret resource for some users.